store/store.go

```diff
@@ -321,34 +321,93 @@ func (s *SnapUbuntuStoreRepository) getPurchasesFromURL(url *url.URL, channel st
 	return purchases, nil
 }
 
-// getPurchases returns user purchases for a specific snap, including in-app.
-func (s *SnapUbuntuStoreRepository) getPurchases(snapID, channel string, auther Authenticator) ([]*purchase, error) {
-	purchasesURL, err := s.purchasesURI.Parse(snapID + "/")
+// decoratePurchases determines if a snap must be bought and sets the MustBuy property accordingly
+func (s *SnapUbuntuStoreRepository) decoratePurchases(info *snap.Info, channel string, auther Authenticator) error {
+	if len(info.Prices) == 0 {
+		// No need to lookup purchases for a free snap
+		return nil
+	}
+
+	if auther == nil {
+		// If we can't authenticate, then the snap must be marked as must buy.
+		info.MustBuy = true
+		return nil
+	}
+
+	purchasesURL, err := s.purchasesURI.Parse(info.SnapID + "/")
 	if err != nil {
-		return nil, err
+		return err
 	}
 
 	q := purchasesURL.Query()
 	q.Set("include_item_purchases", "true")
 	purchasesURL.RawQuery = q.Encode()
 
-	return s.getPurchasesFromURL(purchasesURL, channel, auther)
+	purchases, err := s.getPurchasesFromURL(purchasesURL, channel, auther)
+	if err != nil {
+		info.MustBuy = true
+		return err
+	}
+
+	info.MustBuy = mustBuy(info.Prices, purchases)
+
+	return nil
+}
+
+func setAllNonFreeSnapsAsMustBuy(snaps []*snap.Info) {
+	for _, info := range snaps {
+		if len(info.Prices) != 0 {
+			info.MustBuy = true
+		}
+	}
 }
 
-// getAllPurchases returns all user purchases as a map indexed by snap id.
-func (s *SnapUbuntuStoreRepository) getAllPurchases(channel string, auther Authenticator) (map[string][]*purchase, error) {
+// decorateAllPurchases returns all user purchases as a map indexed by snap id.
+func (s *SnapUbuntuStoreRepository) decorateAllPurchases(snaps []*snap.Info, channel string, auther Authenticator) error {
```

Comment:
Actually, looking above, the single method seems to be entirely a special case of this general one. I'd suggest dropping it altogether and simply tuning this method to tune the URL if len(snaps) == 1. It'll be less logic, less duplication, which means less risk of having bugs due to out-of-date semantics between the two methods.